app/src/app/api/v0/chat/[inventory]/route.ts

```diff
@@ -0,0 +1,277 @@
+import { z } from "zod";
+import { HuggingFaceStream, OpenAIStream, StreamingTextResponse } from "ai";
+import { experimental_buildOpenAssistantPrompt } from "ai/prompts";
+import OpenAI from "openai";
+import { ChatCompletionMessageParam } from "openai/resources/chat/completions";
+import { apiHandler } from "@/util/api";
+import UserService from "@/backend/UserService";
+import { db } from "@/models";
+import { Inventory } from "@/models/Inventory";
+import { Population } from "@/models/Population";
+
+let Hf: HfInference, openai: OpenAI;
+
+function setupModels() {
+  if (!Hf) {
+    Hf = new HfInference(process.env.HUGGINGFACE_API_KEY);
+  }
+  if (!openai) {
+    openai = new OpenAI({
+      apiKey: process.env.OPENAI_API_KEY,
+    });
+  }
+}
+
+// export const runtime = "edge";
+
+const roleSchema = z.enum([
+  "function",
+  "data",
+  "system",
+  "user",
+  "assistant",
+  "tool",
+]);
+const messagesSchema = z.array(
+  z.object({ content: z.string(), role: roleSchema }),
+);
+const chatRequest = z.object({
+  messages: messagesSchema,
+});
+type Messages = z.infer<typeof messagesSchema>;
+
+// Function below is not being referenced/implemented
+
+// async function streamToString(stream: any) {
+//   const reader = stream.getReader();
+//   const textDecoder = new TextDecoder();
+//   let result = "";
+
+//   async function read() {
+//     const { done, value } = await reader.read();
+
+//     if (done) {
+//       return result;
+//     }
+
+//     result += textDecoder.decode(value, { stream: true });
+//     return read();
+//   }
+
+//   return read();
+// }
+
+async function handleHuggingFaceChat(
+  messages: Messages,
+): Promise<StreamingTextResponse> {
+  const response = Hf.textGenerationStream({
+    // The chosen model below is performing much worse compared to GPT3.5
+    // Should be exchanged with a more performant open model from HF that also supports system messages
+    model: "OpenAssistant/oasst-sft-4-pythia-12b-epoch-3.5",
+    inputs: experimental_buildOpenAssistantPrompt(messages),
+    parameters: {
+      max_new_tokens: 200,
+      // @ts-ignore (valid param for OpenAssistant models)
+      typical_p: 0.2,
+      repetition_penalty: 1,
+      truncate: 1000,
+      return_full_text: false,
+    },
+  });
+
+  const stream = HuggingFaceStream(response);
+
+  return new StreamingTextResponse(stream);
+}
+
+async function handleOpenAIChat(
+  messages: Messages,
+): Promise<StreamingTextResponse> {
+  const response = await openai.chat.completions.create({
+    model: "gpt-3.5-turbo",
+    stream: true,
+    messages: messages as any as ChatCompletionMessageParam[],
+  });
+  const stream = OpenAIStream(response);
+  return new StreamingTextResponse(stream);
+}
+
+/**
+ * Creates a system message with context included
+ */
+function create_prompt_template(inventory: Inventory): string {
+  /**
+   * Helper function to extract the latest population value
+   */
+  function find_last_city_population(
```

Comment:
Usually, you don't want the last one for all of these, rather the one that's closest to the selected inventory year (or even better, use interpolation/ extrapolation).

There are already helper methods in the codebase that do this (at least the first part). See getClosestPopulationYear or similar in a util.ts file.

Also, method names in our TS code style are not snake_case but should be camelCase.

Reply:
`findClosestYearToInventory` returns the Population object  that contains the (city) population closest to the given year. This object may or may not contain populated fields for countyPopulation or regionPopulation.

`findClosestYear` returns he Population object back closest to the given year which contains any of the three populations fields. 

None of these two functions return all 3 values closest to the given year (e.g. one entry might only have 2 values, another entry would have the missing 3rd entry).

I will change my file to use `findClosestYearToInventory` for the latest (city) population for now.